### ISSUE TITLE
Karma and chai-enzyme

### ISSUE THREAD

**Issue by mxstbr** [mxstbr/react-boilerplate#119]
We had to remove chai-enzyme to get our test setup with Karma working. This is a shame, because the output by chai-enzyme when a test fails is incredibly useful, much better than the current one.

I want to add chai-enzyme again, but there are a few issues to be resolved which @jbinto is working on.

*See https://github.com/mxstbr/react-boilerplate/issues/79#issuecomment-174238581 (archived) and https://github.com/producthunt/chai-enzyme/issues/16 (active discussion) for more information.*

**Comment by jbinto** [comment]
I'll work on this again now that `chai-enzyme` has been updated to work with plain `require`:

https://github.com/producthunt/chai-enzyme/issues/16#issuecomment-178243404

**Comment by jbinto** [comment]
FWIW, I took another look at this today and I'm still confused.

I am pulling in `karma-chai-enzyme` (which is a naive copy paste of `karma-dirty-chai`) and getting this:

```
Uncaught ReferenceError: require is not defined
  at /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/build/index.js:3
```

**Comment by jbinto** [comment]
Another boilerplate (davezuko/react-redux-starter-kit) seems to have gotten `chai-enzyme` working with Karma:

https://github.com/davezuko/react-redux-starter-kit/commit/311aab2bebd50104f4932ded80568bf24a43c27b

**Comment by jbinto** [comment]
- [x] include `sinon` in `resolve.alias`
- [x] define externals for enzyme (`jsdom`, `cheerio`, ...) [(ref)](https://github.com/davezuko/react-redux-starter-kit/commit/311aab2bebd50104f4932ded80568bf24a43c27b#diff-9351c138c16fc31e777d6b9240b0d32cR47)
- [ ] use ImportsLoader for `sinon` (`require`, `define`) [(ref)](https://github.com/davezuko/react-redux-starter-kit/commit/311aab2bebd50104f4932ded80568bf24a43c27b#diff-9351c138c16fc31e777d6b9240b0d32cR43) *(n.b. not sure why this is needed)*
- [ ] create `test-bundler.js`
- [ ] test-bundler.js: `chai.use` statements [(ref)](https://github.com/davezuko/react-redux-starter-kit/commit/311aab2bebd50104f4932ded80568bf24a43c27b#diff-74e9d8a8f3f5500f7401330987221aafR10)
- [ ] test-bundler.js: require all tests [(ref)](https://github.com/davezuko/react-redux-starter-kit/commit/311aab2bebd50104f4932ded80568bf24a43c27b#diff-74e9d8a8f3f5500f7401330987221aafR23)
- [ ] karma.conf: preprocess `test-bundler` instead of `**` [(ref)](https://github.com/davezuko/react-redux-starter-kit/commit/311aab2bebd50104f4932ded80568bf24a43c27b#diff-9351c138c16fc31e777d6b9240b0d32cL22)
- [ ] restore `chai-enzyme`-style assertions (ref 4ac051b4c454f5ae34fc6d15f08ce5890f1d63c9)

**Comment by jbinto** [comment]
OK, I'm officially lost.

https://github.com/jbinto/react-boilerplate/commit/55f296b74bcaf5f83a10e9f5012a9f5958cb272c

```sh
npm run test 2>&1 | grep 'Module parse'
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/attr.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/checked.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/blank.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/className.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/contain.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/data.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/descendants.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/disabled.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/empty.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/exist.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/html.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/id.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/inspect.test.js Line 1: Unexpected token
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/match.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/null.test.js Line 7: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/present.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/prop.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/ref.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/selected.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/state.test.js Line 9: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/style.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/tagName.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/text.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/test/value.test.js Line 4: Unexpected token <
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/CHANGELOG.md Line 1: Unexpected token ILLEGAL
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/LICENSE Line 1: Unexpected identifier
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/Readme.md Line 1: Unexpected token ILLEGAL
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/binding.gyp Line 2: Unexpected token :
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/fsevents.cc Line 6: Unexpected token ILLEGAL
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/src/constants.cc Line 7: Unexpected token ILLEGAL
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/src/async.cc Line 7: Unexpected token {
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/src/locking.cc Line 6: Unexpected token :
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/src/methods.cc Line 6: Unexpected token :
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/src/storage.cc Line 6: Unexpected identifier
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/src/thread.cc Line 7: Unexpected token ILLEGAL
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/lib/binding/Release/node-v47-darwin-x64/fse.node Line 1: Unexpected token ILLEGAL
Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/node-pre-gyp/lib/util/nw-pre-gyp/index.html Line 1: Unexpected token <
```

```sh
npm run test 2>&1 | grep 'Module not found'
Module not found: Error: Cannot resolve module 'node-gyp' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/node-pre-gyp/lib/util
Module not found: Error: Cannot resolve module 'npm' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/node-pre-gyp/lib/util
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/base64url/test
Module not found: Error: Cannot resolve module 'should' in /Users/jbinto/dev/react-boilerplate/node_modules/fastparse/test
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/jwa/node_modules/base64url/test
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/jwa/test
Module not found: Error: Cannot resolve module 'tape' in /Users/jbinto/dev/react-boilerplate/node_modules/jws/test
Module not found: Error: Cannot resolve module 'should' in /Users/jbinto/dev/react-boilerplate/node_modules/watchpack/test
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/google-auth-library/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/google-auth-library/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/googleapis/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/googleapis/node_modules/asn1/tst/ber
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/test
Module not found: Error: Cannot resolve module 'tap' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/test
Module not found: Error: Cannot resolve module 'aws-sdk' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/node-pre-gyp/lib
Module not found: Error: Cannot resolve module 'aws-sdk' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/node-pre-gyp/lib
Module not found: Error: Cannot resolve module 'aws-sdk' in /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/node_modules/node-pre-gyp/lib
```

**Comment by jbinto** [comment]
Looking really closely at davezuko/react-redux-starter-kit and I just can't see what I'm doing differently.

**Comment by jbinto** [comment]
WTF is it doing trying to import a `.md` file?

`Module parse failed: /Users/jbinto/dev/react-boilerplate/node_modules/fsevents/CHANGELOG.md Line 1: Unexpected token ILLEGAL`

**Comment by jbinto** [comment]
Aha!

https://github.com/jbinto/react-boilerplate/commit/55f296b74bcaf5f83a10e9f5012a9f5958cb272c#diff-507b7b6eb038b4024f4fd14f8528c2feR20

I'm pulling in `*/**/*.test.js`, **including inside `node_modules`**.

**Comment by jbinto** [comment]
The other boilerplate has a solid separation between `src` and `test`. 

The way this repo is set up makes it a little more difficult to make a single glob that describes the tests we want to target.

Also see #149

**Comment by jbinto** [comment]
https://github.com/jbinto/react-boilerplate/commit/551858441275d8efbf3b40762f29f4871b8f30fd

OK, now I'm on the right track. Gotta take a break though, I'm in a coffee shop and my laptop battery is about to die.

**Comment by jbinto** [comment]
Perspective:

Before #117, the test setup was mocha+jsdom executing in a Node environment.

After #117, the test setup is mocha+karma executing in Chrome. I remember arguing that using the same JS environment as production along with a real DOM for tests makes me feel safer.

After introducing `enzyme`, the test setup becomes mocha+karma+jsdom executing in Chrome, since `enzyme` has a hard dependency on `jsdom`. So we have a real non-Node JS environment, a real DOM, but now we're executing `jsdom` inside Chrome.

Gonna want to keep a close eye on https://github.com/airbnb/enzyme/pull/159 which would update enzyme to use a real DOM if it can find one.

**Comment by jbinto** [comment]
Fixed in #150.

**Comment by mxstbr** [comment]
Thanks for digging that deep into this and getting it back, amazing work!!

----

@jongold we got chai-enzyme back!

@vesln we got chai-enzyme working here, you might want to check
